### PR TITLE
[3.4.2] Feature: check tenant entity for rule nodes

### PR DESCRIPTION
**Pull request by YuriyLytvynchuk**
## Pull Request description
issue: https://github.com/thingsboard/thingsboard/issues/7393

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [x] PR name contains fix version. For example, "[3.3.4] Hotfix of some UI component" or "[3.4] New Super Feature".
- [x] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [ ] Similar PR is opened for PE version to simplify merge. Required for internal contributors only.
  
## Front-End feature checklist

- [x] Screenshots with affected component(s) are added. The best option is to provide 2 screens: before and after changes;
- [x] If you change the widget or other API, ensure it is backward-compatible or upgrade script is present.
- [x] Ensure new API is documented [here](https://github.com/thingsboard/thingsboard-ui-help)

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.

### CODE REVIEW

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/debug/TbMsgGeneratorNode.java [hunk 1]
```diff
@@ -80,6 +80,7 @@ public void init(TbContext ctx, TbNodeConfiguration configuration) throws TbNode
         } else {
             originatorId = ctx.getSelfId();
         }
+        ctx.checkTenantEntity(originatorId);
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/debug/TbMsgGeneratorNode.java:83]: move this validation into if statement

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/filter/TbCheckRelationNode.java [hunk 2]
```diff
@@ -59,6 +59,7 @@ public class TbCheckRelationNode implements TbNode {
     @Override
     public void init(TbContext ctx, TbNodeConfiguration configuration) throws TbNodeException {
         this.config = TbNodeUtils.convert(configuration, TbCheckRelationNodeConfiguration.class);
+        ctx.checkTenantEntity(EntityIdFactory.getByTypeAndId(config.getEntityType(), config.getEntityId()));
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/filter/TbCheckRelationNode.java:62]: Why do we need to validate this if isCheckForSingleEntity is set to false?

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/util/EntitiesTenantIdAsyncLoader.java [hunk 3]
```diff
@@ -57,10 +63,42 @@ public static ListenableFuture<TenantId> findEntityIdAsync(TbContext ctx, Entity
         }
     }
 
+    public static ListenableFuture<TenantId> findTenantIdByEntityId(TbContext ctx, EntityId entityId) {
```
[the remark this answers is not in the document]
Comment (ShvaykaD) [reply rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/util/EntitiesTenantIdAsyncLoader.java:66]: Remove deprecated annotation and description.

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/util/EntitiesTenantIdAsyncLoader.java [hunk 4]
```diff
@@ -57,10 +63,42 @@ public static ListenableFuture<TenantId> findEntityIdAsync(TbContext ctx, Entity
...
+        }
+    }
+
     private static <T extends HasTenantId> ListenableFuture<TenantId> getTenantAsync(ListenableFuture<T> future) {
-        return Futures.transformAsync(future, in -> {
-            return in != null ? Futures.immediateFuture(in.getTenantId())
-                    : Futures.immediateFuture(null);
-        }, MoreExecutors.directExecutor());
+        return Futures.transformAsync(future, in -> in != null ? Futures.immediateFuture(in.getTenantId())
+                : Futures.immediateFuture(null), MoreExecutors.directExecutor());
```
[the remark this answers is not in the document]
Comment (ShvaykaD) [reply rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/util/EntitiesTenantIdAsyncLoader.java:102]: Also, please rollback formatting changes.

**Review by ShvaykaD — APPROVED** [review]

**Comment by YevhenBondarenko** [comment]
Plz, add PR for the PE.

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/util/EntitiesTenantIdAsyncLoader.java [hunk 5]
```diff
@@ -17,50 +17,121 @@
...
+                hasTenantId = Futures.immediateFuture(ctx.getDeviceProfileCache().get(tenantId, new DeviceProfileId(id)));
+                break;
+            case WIDGET_TYPE:
+                hasTenantId = Futures.immediateFuture(ctx.getWidgetTypeService().findWidgetTypeById(tenantId, new WidgetTypeId(id)));
+                break;
+            case WIDGETS_BUNDLE:
+                hasTenantId = Futures.immediateFuture(ctx.getWidgetBundleService().findWidgetsBundleById(tenantId, new WidgetsBundleId(id)));
+                break;
+            case RPC:
+                hasTenantId = ctx.getRpcService().findRpcByIdAsync(ctx.getTenantId(), new RpcId(id));
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/util/EntitiesTenantIdAsyncLoader.java:103]: replace with tenantId

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/util/EntitiesTenantIdAsyncLoader.java [hunk 6]
```diff
@@ -17,50 +17,121 @@
...
+                hasTenantId = Futures.immediateFuture(ctx.getRuleEngineApiUsageStateService().findApiUsageStateById(tenantId, new ApiUsageStateId(id)));
+                break;
+            case TB_RESOURCE:
+                hasTenantId = ctx.getResourceService().findResourceInfoByIdAsync(tenantId, new TbResourceId(id));
+                break;
+            case RULE_NODE:
+                hasTenantId = null;
+                RuleNode ruleNode = ctx.getRuleChainService().findRuleNodeById(tenantId, new RuleNodeId(id));
+                if (ruleNode != null) {
+                    return Futures.immediateFuture(tenantId);
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/util/EntitiesTenantIdAsyncLoader.java:118]: please fetch the rule chain if from rule node and then fetch id of tenant from the rule chain

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/util/EntitiesTenantIdAsyncLoader.java [hunk 7]
```diff
@@ -17,50 +17,121 @@
...
+                }
+                break;
+            case TENANT_PROFILE:
+                hasTenantId = null;
+                if (ctx.getTenantProfile() == ctx.getTenantProfile(new TenantId(id))) {
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/util/EntitiesTenantIdAsyncLoader.java:123]: check originator id with  id of profile from ctx.getTenantProfile()

**Pull request merged**